### ISSUE TITLE
TimeoutException in ConnectionPool.handlerFromSharedCache

### ISSUE THREAD

**Issue by hmlnarik** [keycloak/keycloak#19717]
After fixing #16448, [some users](https://github.com/keycloak/keycloak/issues/16448#issuecomment-1504734541) still experience `TimeoutException` in `ConnectionPool.handlerFromSharedCache`.

This is a tracker issue for https://issues.redhat.com/browse/AG-204. A reference for Quarkus issue may be added once known.

This issue is currently blocked on the aforementioned `AG-204` issue.

```
C:\Users\***\keycloak-21.0.2>bin\kc start-dev
2023-04-13 10:16:57,717 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: Base URL: <unset>, Hostname: <request>, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin URL: <unset>, Admin: <request>, Port: -1, Proxied: false
2023-04-13 10:16:59,194 WARN  [io.quarkus.agroal.runtime.DataSources] (main) Datasource <default> enables XA but transaction recovery is not enabled. Please enable transaction recovery by setting quarkus.transaction-manager.enable-recovery=true, otherwise data may be lost if the application is terminated abruptly
2023-04-13 10:17:01,858 WARN  [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000569: Unable to persist Infinispan internal caches as no global state enabled
2023-04-13 10:17:01,890 WARN  [org.infinispan.PERSISTENCE] (keycloak-cache-init) ISPN000554: jboss-marshalling is deprecated and planned for removal
2023-04-13 10:17:01,921 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
2023-04-13 10:17:04,366 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: keycloak-default) HHH000342: Could not obtain connection to query metadata: java.sql.SQLException: Acquisition timeout while waiting for new connection
        at io.agroal.pool.ConnectionPool.handlerFromSharedCache(ConnectionPool.java:321)
        at io.agroal.pool.ConnectionPool.getConnection(ConnectionPool.java:249)
        at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
        at io.quarkus.hibernate.orm.runtime.customized.QuarkusConnectionProvider.getConnection(QuarkusConnectionProvider.java:23)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$ConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:181)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:68)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:35)
        at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:101)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:263)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:237)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
        at org.hibernate.engine.jdbc.internal.JdbcServicesImpl.configure(JdbcServicesImpl.java:51)
        at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.configureService(StandardServiceRegistryImpl.java:107)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:246)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
        at org.hibernate.boot.internal.SessionFactoryOptionsBuilder.<init>(SessionFactoryOptionsBuilder.java:272)
        at io.quarkus.hibernate.orm.runtime.recording.PrevalidatedQuarkusMetadata.buildSessionFactoryOptionsBuilder(PrevalidatedQuarkusMetadata.java:68)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:72)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:72)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:165)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:66)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.TimeoutException
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
        at io.agroal.pool.ConnectionPool.handlerFromSharedCache(ConnectionPool.java:298)
        ... 23 more

2023-04-13 10:17:06,653 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': IO Exception: "C:/Users/***/keycloak-21.0.2/data/h2/keycloakdb.mv.db" [90028-214]
2023-04-13 10:17:06,668 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': Error opening database: "Sleep interrupted" [8000-214]
2023-04-13 10:17:06,842 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (development) mode
2023-04-13 10:17:06,842 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to obtain JDBC connection
2023-04-13 10:17:06,842 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Error opening database: "Sleep interrupted" [8000-214]
2023-04-13 10:17:06,842 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: sleep interrupted
2023-04-13 10:17:06,842 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
```

_Originally posted by @luke-sensei in https://github.com/keycloak/keycloak/issues/16448#issuecomment-1506550481_

**Comment by sschu** [comment]
@hmlnarik I am not sure this is actually an agroal issue. To me, it looks more like H2 indeed has a problem connecting to local files sometimes and runs into a timeout - which is definitely strange.

**Comment by hmlnarik** [comment]
@sschu I agree and I am not sure either, since I was not able to create a reproducer yet. Still the worrisome aspect is that [reportedly](https://github.com/keycloak/keycloak/issues/16448#issuecomment-1504734541) downgrading agroal to 1.15 fixes this issue.

**Comment by SXI-XLayer** [comment]
> @sschu I agree and I am not sure either, since I was not able to create a reproducer yet. Still the worrisome aspect is that [reportedly](https://github.com/keycloak/keycloak/issues/16448#issuecomment-1504734541) downgrading agroal to 1.15 fixes this issue.

I can confirm.  I have just tried the entire day to get this working and stumbled on this thead in desparation.  Here is what I have.

Windows 11
Keycloak v21.1.0
OpenJDK v20.0.1

io.agroal.agroal-api-1.17.jar
io.agroal.agroal-narayana-1.17.jar
io.agroal.agroal-pool-1.17.jar

When I run `bin\kc.bat start-dev` I get the following:

> 2023-04-20 19:22:21,840 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: Base URL: <unset>, Hostname: <request>, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin URL: <unset>, Admin: <request>, Port: -1, Proxied: false
> 2023-04-20 19:22:22,851 WARN  [io.quarkus.agroal.runtime.DataSources] (main) Datasource <default> enables XA but transaction recovery is not enabled. Please enable transaction recovery by setting quarkus.transaction-manager.enable-recovery=true, otherwise data may be lost if the application is terminated abruptly
> 2023-04-20 19:22:23,483 WARN  [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000569: Unable to persist Infinispan internal caches as no global state enabled
> 2023-04-20 19:22:23,493 WARN  [org.infinispan.PERSISTENCE] (keycloak-cache-init) ISPN000554: jboss-marshalling is deprecated and planned for removal
> 2023-04-20 19:22:23,511 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
> 2023-04-20 19:22:28,044 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: keycloak-default) HHH000342: Could not obtain connection to query metadata: java.sql.SQLException: Acquisition timeout while waiting for new connection
>         at io.agroal.pool.ConnectionPool.handlerFromSharedCache(ConnectionPool.java:321)
>         at io.agroal.pool.ConnectionPool.getConnection(ConnectionPool.java:249)
>         at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
>         at io.quarkus.hibernate.orm.runtime.customized.QuarkusConnectionProvider.getConnection(QuarkusConnectionProvider.java:23)
>         at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$ConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:181)
>         at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:68)
>         at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:35)
>         at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:101)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:263)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:237)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
>         at org.hibernate.engine.jdbc.internal.JdbcServicesImpl.configure(JdbcServicesImpl.java:51)
>         at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.configureService(StandardServiceRegistryImpl.java:107)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:246)
>         at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
>         at org.hibernate.boot.internal.SessionFactoryOptionsBuilder.<init>(SessionFactoryOptionsBuilder.java:272)
>         at io.quarkus.hibernate.orm.runtime.recording.PrevalidatedQuarkusMetadata.buildSessionFactoryOptionsBuilder(PrevalidatedQuarkusMetadata.java:68)
>         at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:72)
>         at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:72)
>         at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
>         at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
>         at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:165)
>         at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:66)
>         at java.base/java.lang.Thread.run(Thread.java:1623)
> Caused by: java.util.concurrent.TimeoutException
>         at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
>         at io.agroal.pool.ConnectionPool.handlerFromSharedCache(ConnectionPool.java:298)
>         ... 23 more
> 
> 2023-04-20 19:22:28,185 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': IO Exception: "D:/XLayer_Utils/XIAM/data/h2/keycloakdb.mv.db" [90028-214]
> 2023-04-20 19:22:33,341 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': IO Exception: "D:/XLayer_Utils/XIAM/data/h2/keycloakdb.mv.db" [90028-214]
> 2023-04-20 19:22:33,481 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (development) mode
> 2023-04-20 19:22:33,482 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to obtain JDBC connection
> 2023-04-20 19:22:33,482 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Acquisition timeout while waiting for new connection
> 2023-04-20 19:22:33,482 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.

I also tried with the newer versions:
agroal-api-2.1.jar
agroal-narayana-2.1.jar
agroal-pool-2.1.jar

but as you can imagine - that was extremely wishful thinking.  If you would like I can test some of the other versions of the agroal libs.

Ciao

**Comment by SXI-XLayer** [comment]
> @hmlnarik I am not sure this is actually an agroal issue. To me, it looks more like H2 indeed has a problem connecting to local files sometimes and runs into a timeout - which is definitely strange.

Just to shed a little more light on this if it helps.  I am getting this issue with H2 and with MariaDB runnign in a docker container.  The downgrade of the agroal libs got it working again.

**Comment by kami619** [comment]
@hmlnarik I am getting a similar error with postgresql and legacy store on the openshift provisioning., not sure if its coming the from same use case or a different one, but searching for `DefaultHostnameProvider` keyword got me here.

And the thing with this is, you are able to login into the admin account, however as a regular use you are not able to use the account-console or custom clients like we use in keycloak-benchmark, they fail with a 404 on the client side.

cc: @ahus1 

```
$ kubectl logs -n keycloak keycloak-0
Appending additional Java properties to JAVA_OPTS: -Xms1024m -Xmx2048m -XX:MetaspaceSize=1024m -XX:MaxMetaspaceSize=1024m -Djava.security.egd=file:/dev/urandom -Dnetworkaddress.cache.ttl=10 -XX:+ExitOnOutOfMemoryError -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:8787

Listening for transport dt_socket at address: 8787
Changes detected in configuration. Updating the server image.
Updating the configuration and installing your custom providers, if any. Please wait.
2023-04-21 19:58:49,912 INFO  [io.quarkus.deployment.QuarkusAugmentor] (main) Quarkus augmentation completed in 7470ms
Server configuration updated and persisted. Run the following command to review the configuration:

	kc.sh show-config

Next time you run the server, just run:

	kc.sh start --optimized 

Listening for transport dt_socket at address: 8787
{"timestamp":"2023-04-21T19:58:51.274Z","sequence":28,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider","level":"INFO","message":"Hostname settings: Base URL: <unset>, Hostname: keycloak.apps.d17-h28-000-r650.example.com, Strict HTTPS: true, Path: <request>, Strict BackChannel: false, Admin URL: <unset>, Admin: <request>, Port: -1, Proxied: true","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.254Z","sequence":8396,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.quarkus.agroal.runtime.DataSources","level":"WARN","message":"Datasource <default> enables XA but transaction recovery is not enabled. Please enable transaction recovery by setting quarkus.transaction-manager.enable-recovery=true, otherwise data may be lost if the application is terminated abruptly","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.382Z","sequence":8473,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.agroal.pool","level":"WARN","message":"Datasource '<default>': FATAL: the database system is shutting down","threadName":"agroal-11","threadId":20,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.394Z","sequence":8474,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.agroal.pool","level":"WARN","message":"Datasource '<default>': FATAL: the database system is shutting down","threadName":"agroal-11","threadId":20,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.422Z","sequence":8475,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.agroal.pool","level":"WARN","message":"Datasource '<default>': FATAL: the database system is shutting down","threadName":"agroal-11","threadId":20,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.809Z","sequence":8476,"loggerClassName":"org.infinispan.server.core.logging.Log_$logger","loggerName":"org.infinispan.SERVER","level":"INFO","message":"ISPN005054: Native IOUring transport not available, using NIO instead: io.netty.incubator.channel.uring.IOUring","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.872Z","sequence":8477,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.quarkus.vertx.http.runtime.VertxHttpRecorder","level":"WARN","message":"The X-Forwarded-* and Forwarded headers will be considered when determining the proxy address. This configuration can cause a security issue as clients can forge requests and send a forwarded header that is not overwritten by the proxy. Please consider use one of these headers just to forward the proxy address in requests.","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.883Z","sequence":8478,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.agroal.pool","level":"WARN","message":"Datasource '<default>': FATAL: the database system is shutting down","threadName":"agroal-11","threadId":20,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.9Z","sequence":8479,"loggerClassName":"org.infinispan.util.logging.Log_$logger","loggerName":"org.infinispan.PERSISTENCE","level":"WARN","message":"ISPN000554: jboss-marshalling is deprecated and planned for removal","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.929Z","sequence":8480,"loggerClassName":"org.infinispan.util.logging.Log_$logger","loggerName":"org.infinispan.CONFIG","level":"WARN","message":"ISPN000569: Unable to persist Infinispan internal caches as no global state enabled","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:52.944Z","sequence":8481,"loggerClassName":"org.infinispan.util.logging.Log_$logger","loggerName":"org.infinispan.CONTAINER","level":"INFO","message":"ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:53.3Z","sequence":8482,"loggerClassName":"org.infinispan.util.logging.Log_$logger","loggerName":"org.infinispan.CLUSTER","level":"INFO","message":"ISPN000078: Starting JGroups channel `ISPN`","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:53.304Z","sequence":8483,"loggerClassName":"org.slf4j.impl.Slf4jLogger","loggerName":"org.jgroups.JChannel","level":"INFO","message":"local_addr: 9ab108a4-e735-45c9-b083-3b780fced2f5, name: keycloak-0-46636","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:53.31Z","sequence":8484,"loggerClassName":"org.slf4j.impl.Slf4jLogger","loggerName":"org.jgroups.protocols.FD_SOCK2","level":"INFO","message":"server listening on *.57800","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:53.8Z","sequence":8485,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.agroal.pool","level":"WARN","message":"Datasource '<default>': FATAL: the database system is shutting down","threadName":"agroal-11","threadId":20,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:53.828Z","sequence":8486,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider","level":"INFO","message":"Initializing database schema. Using changelog META-INF/jpa-changelog-master.xml","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:55.313Z","sequence":8487,"loggerClassName":"org.slf4j.impl.Slf4jLogger","loggerName":"org.jgroups.protocols.pbcast.GMS","level":"INFO","message":"keycloak-0-46636: no members discovered after 2002 ms: creating cluster as coordinator","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:55.321Z","sequence":8488,"loggerClassName":"org.infinispan.util.logging.Log_$logger","loggerName":"org.infinispan.CLUSTER","level":"INFO","message":"ISPN000094: Received new cluster view for channel ISPN: [keycloak-0-46636|0] (1) [keycloak-0-46636]","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:55.326Z","sequence":8489,"loggerClassName":"org.infinispan.util.logging.Log_$logger","loggerName":"org.infinispan.CLUSTER","level":"INFO","message":"ISPN000079: Channel `ISPN` local address is `keycloak-0-46636`, physical addresses are `[10.128.1.230:7800]`","threadName":"keycloak-cache-init","threadId":23,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:55.679Z","sequence":8490,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory","level":"INFO","message":"Node name: keycloak-0-46636, Site name: null","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:55.754Z","sequence":8491,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.broker.provider.AbstractIdentityProviderMapper","level":"INFO","message":"Registering class org.keycloak.broker.provider.mappersync.ConfigSyncEventListener","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:55.767Z","sequence":8492,"loggerClassName":"org.keycloak.services.ServicesLogger_$logger","loggerName":"org.keycloak.services","level":"INFO","message":"KC-SERVICES0050: Initializing master realm","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:56.968Z","sequence":8493,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.quarkus","level":"INFO","message":"Keycloak 999.0.0-SNAPSHOT on JVM (powered by Quarkus 2.13.7.Final) started in 6.599s. Listening on: https://0.0.0.0:8443","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:56.968Z","sequence":8494,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.quarkus","level":"INFO","message":"Profile prod activated. ","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:56.968Z","sequence":8495,"loggerClassName":"org.jboss.logging.Logger","loggerName":"io.quarkus","level":"INFO","message":"Installed features: [agroal, cdi, hibernate-orm, jdbc-h2, jdbc-mariadb, jdbc-mssql, jdbc-mysql, jdbc-oracle, jdbc-postgresql, keycloak, logging-gelf, micrometer, narayana-jta, reactive-routes, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-health, vertx]","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}
{"timestamp":"2023-04-21T19:58:57.105Z","sequence":8496,"loggerClassName":"org.keycloak.services.ServicesLogger_$logger","loggerName":"org.keycloak.services","level":"INFO","message":"KC-SERVICES0009: Added user 'admin' to realm 'master'","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}

```

**Comment by ahus1** [comment]
@kami619 - I'm seeing "the database system is shutting down" which is valid error message after a system-under-test reset when the Keycloak restart  was faster than the PostgreSQL shutdown. The global lock might then wait for some time for the database to come up again. I don't see a TimeoutException in your log snippet, therefore I consider it unrelated.

**Comment by kami619** [comment]
> @kami619 - I'm seeing "the database system is shutting down" which is valid error message after a system-under-test reset when the Keycloak restart was faster than the PostgreSQL shutdown. The global lock might then wait for some time for the database to come up again. I don't see a TimeoutException in your log snippet, therefore I consider it unrelated.

thank you for the clarification Alexander, appreciate it

**Comment by ssonparo** [comment]
@hmlnarik and @sschu  : I don' think this issue has anything to do with agroal-pool-*.*.jar . 
Configuration I am using .
Keycloak version 21.0.1.
openjdk/alpine:***-image-jdk-17
I have used H2 internal DB ,Postgres db with docker container . This combination works good on Docker Desktop on local machine.
Once I move it to kubernetes cluster  centos using docker container . It creates problem .

**Comment by sschu** [comment]
@ssonparo What exactly is the problem? Do you always run into a timeout or only sometimes? If for some reason your external database cannot be reached, that might also show up with this error. The original issue here was that this happens with H2 when accessing a database in a local file - and this should actually never time out.

**Comment by ssonparo** [comment]
@sschu : Thank you for response . 
CONFIGURATION : "
I run into this issue every time, whenever  I am trying to start keycloak in Dev mode . I am creating docker image and deploying keycloak within the kubernetes cluster pods ." 
The same setup works fine if I run keycloak in dev mode on local machine using Docker Desktop (NO kubernetes cluster pods ).Integration with H2 database and Postgres database as well . This works fine without any exception .
I understand that this issue is for H2 database ,I am following this thread  and another  as well (https://github.com/keycloak/keycloak/issues/16448)but once I try to deploy using above mention configuration , it breaks with timeout exception .
The reason I am saying it is not related to agoral related  jars is because .The database connection it is up and running.

**Comment by sschu** [comment]
@ssonparo But to me this sounds like something is wrong in your network setup if you are not able to connect every time you try. You should probably do a bit of digging on the network level of your database is reachable as you expect at all.

**Comment by kandadeepak** [comment]
Hi All,

I'm using ubuntu 20.04 with docker for keycloak along with postgress:latest.

Docker file 
FROM quay.io/keycloak/keycloak:latest as builder

# Enable health and metrics support
ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

# Configure a database vendor
ENV KC_DB=postgres

WORKDIR /opt/keycloak
# for demonstration purposes only, please make sure to use proper certificates in production instead
RUN keytool -genkeypair -storepass password -storetype PKCS12 -keyalg RSA -keysize 2048 -dname "CN=server" -alias server -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore conf/server.keystore
RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:latest
COPY --from=builder /opt/keycloak/ /opt/keycloak/

# change these values to point to a running postgres instance
ENV KC_DB=postgres
ENV KC_DB_URL=jdbc:postgresql://localhost:32769/keycloak
ENV KC_DB_USERNAME=keycloak
ENV KC_DB_PASSWORD=password
ENV KC_HOSTNAME=localhost
ENV KC_HOSTNAME_STRICT=false
ENV KC_HOSTNAME_STRICT_HTTPS=false
ENV KC_HTTP_ENABLED=true
ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]




docker run --name mykeycloak -p 8443:8443         -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=xxxx         mykeycloak         start --optimized

2023-05-05 16:35:12,975 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: Base URL: <unset>, Hostname: localhost, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin URL: <unset>, Admin: <request>, Port: -1, Proxied: false
2023-05-05 16:35:14,482 WARN  [io.quarkus.agroal.runtime.DataSources] (main) Datasource <default> enables XA but transaction recovery is not enabled. Please enable transaction recovery by setting quarkus.transaction-manager.enable-recovery=true, otherwise data may be lost if the application is terminated abruptly
2023-05-05 16:35:15,082 INFO  [org.infinispan.SERVER] (keycloak-cache-init) ISPN005054: Native IOUring transport not available, using NIO instead: io.netty.incubator.channel.uring.IOUring
2023-05-05 16:35:15,176 WARN  [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000569: Unable to persist Infinispan internal caches as no global state enabled
2023-05-05 16:35:15,207 WARN  [org.infinispan.PERSISTENCE] (keycloak-cache-init) ISPN000554: jboss-marshalling is deprecated and planned for removal
2023-05-05 16:35:15,223 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
2023-05-05 16:35:15,564 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000088: Unable to use any JGroups configuration mechanisms provided in properties {}. Using default JGroups configuration!
2023-05-05 16:35:15,672 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000078: Starting JGroups channel `ISPN`
2023-05-05 16:35:15,681 INFO  [org.jgroups.JChannel] (keycloak-cache-init) local_addr: 2ec0b3c6-8321-4fd5-8e73-693d3825612e, name: b0dbc732a071-65299
2023-05-05 16:35:15,690 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the send buffer of socket MulticastSocket was set to 1MB, but the OS only allocated 212.99KB
2023-05-05 16:35:15,691 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the receive buffer of socket MulticastSocket was set to 20MB, but the OS only allocated 212.99KB
2023-05-05 16:35:15,691 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the send buffer of socket MulticastSocket was set to 1MB, but the OS only allocated 212.99KB
2023-05-05 16:35:15,692 WARN  [org.jgroups.protocols.UDP] (keycloak-cache-init) JGRP000015: the receive buffer of socket MulticastSocket was set to 25MB, but the OS only allocated 212.99KB
2023-05-05 16:35:15,702 INFO  [org.jgroups.protocols.FD_SOCK2] (keycloak-cache-init) server listening on *.26913
2023-05-05 16:35:17,707 INFO  [org.jgroups.protocols.pbcast.GMS] (keycloak-cache-init) b0dbc732a071-65299: no members discovered after 2001 ms: creating cluster as coordinator
2023-05-05 16:35:17,716 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000094: Received new cluster view for channel ISPN: [b0dbc732a071-65299|0] (1) [b0dbc732a071-65299]
2023-05-05 16:35:17,727 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000079: Channel `ISPN` local address is `b0dbc732a071-65299`, physical addresses are `[172.17.0.2:42449]`
2023-05-05 16:35:19,693 WARN  [org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator] (JPA Startup Thread: keycloak-default) HHH000342: Could not obtain connection to query metadata: java.sql.SQLException: Acquisition timeout while waiting for new connection
        at io.agroal.pool.ConnectionPool.handlerFromSharedCache(ConnectionPool.java:321)
        at io.agroal.pool.ConnectionPool.getConnection(ConnectionPool.java:249)
        at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
        at io.quarkus.hibernate.orm.runtime.customized.QuarkusConnectionProvider.getConnection(QuarkusConnectionProvider.java:23)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$ConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:181)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:68)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:35)
        at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:101)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:263)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:237)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
        at org.hibernate.engine.jdbc.internal.JdbcServicesImpl.configure(JdbcServicesImpl.java:51)
        at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.configureService(StandardServiceRegistryImpl.java:107)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:246)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:214)
        at org.hibernate.boot.internal.SessionFactoryOptionsBuilder.<init>(SessionFactoryOptionsBuilder.java:272)
        at io.quarkus.hibernate.orm.runtime.recording.PrevalidatedQuarkusMetadata.buildSessionFactoryOptionsBuilder(PrevalidatedQuarkusMetadata.java:68)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:72)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:72)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
        at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:165)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:66)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.TimeoutException
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
        at io.agroal.pool.ConnectionPool.handlerFromSharedCache(ConnectionPool.java:298)
        ... 23 more

2023-05-05 16:35:24,773 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': The connection attempt failed.
2023-05-05 16:35:24,778 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': The connection attempt failed.
2023-05-05 16:35:24,817 INFO  [org.infinispan.CLUSTER] (main) ISPN000080: Disconnecting JGroups channel `ISPN`
2023-05-05 16:35:24,864 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2023-05-05 16:35:24,865 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to obtain JDBC connection
2023-05-05 16:35:24,865 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: The connection attempt failed.
2023-05-05 16:35:24,865 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: postgres
2023-05-05 16:35:24,865 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option.

**Comment by ThomasSteinkamp** [comment]
Hi, 
i hope this helps shedding some light on the root cause: This seems to be a network configuration problem. For us, keycloak works when started on a machine without VPN , but does NOT work when having our company VPN active.

With VPN we are getting the exact same callstack as noted on top of this issue. We are using Keycloak 20.0.3 with openjdk-19-0.2, on Windows 10 21H2. (Sorry, i am reluctant to give out the make of VPN we are using, in order to not disclose details of our company network..) 

We have no clue yet, what the difference may be. I inspected the windows event logs to no avail.

**Comment by sschu** [comment]
@ThomasSteinkamp So having a VPN has an effect even if you create a local database? Thats awkward.  @SXI-XLayer do you also use a VPN?

**Comment by ThomasSteinkamp** [comment]
@sschu Yes. My suspicion (only guessing, i don't know about the internals of H2 or JDBC) is that still some internal tcp connection is being used (i can see that the java process has more listening ports open apart from 8080) and for some reason a connection to the wrong adapter/ip (the physical one instead of the 'virtual' adapter established by VPN) is attempted. But that's only guesswork. 
(I just googled again. I found this http://www.h2database.com/html/advanced.html#server_bind_address link, which seems to  support my theories. It seems that although H2 is a 'local file based database', it is using some network connections)

**Comment by hmlnarik** [comment]
@ThomasSteinkamp Wild guess - do both VPN and physical adapters have addresses of the same network protocols? In other words, is by any chance e.g. IPv6 address present on one interface and not present on another?

**Comment by jgoelen** [comment]
We are running Keycloak version 21.1.1 (image quay.io/keycloak/keycloak:21.1.1)on our staging environment using a PostgreSQL db on a K8S cluster and we same exceptions as described in this issue.

In production we are running Keycloak 19.0.2 using a similar setup without a problem.

**Comment by sschu** [comment]
@ThomasSteinkamp Did your colleague use the Quarkus or the Wildfly version of 19.0.3? Because the Wildfly version uses a completely different connection pooling component and cannot show this error message.

**Comment by sschu** [comment]
But still, did you pick the Quarkus or the Wildfly distribution from https://www.keycloak.org/archive/downloads-19.0.3.html?

**Comment by wtell400** [comment]
VPN is the issue for me here as well.

- VPN on - TimeoutException
- VPN off - Starts like normal.
- VPN on - TimeoutException
- VPN off - Starts like normal. 

Keycloak 21.0.2 in production mode, Java 17 on Windows 11 Pro 22H2. Using Windows VPN (L2TP/IPsec with preshared key) connected via ethernet cable.

**Comment by dwaynebailey** [comment]
Not sure this counts as reliably reproducible, I'm not sure I can easily reduce it to something someone can copy.  But it does point to the VPN like issue being a potential root cause, rather then what I first thought of as a red herring.

I have the following setup:
* EKS cluster
* Postgres RDS database for Keycloak 
* Custom Keycloak 20 image (some custom modules and DB config)
* Helm chart holding all of this (includes secrets, Service Accounts, etc)
* AWS Security Groups for Pods https://docs.aws.amazon.com/eks/latest/userguide/security-groups-for-pods.html - which create ENI for each pod that need it
* Security group controls in place to allow the Keycloak pods to talk to RDS
* Cluster config and deployment

The issue has happened in single instances and in cluster config.

Two ways that are mostly reliable to get the system up have been:
1. Deploy everything but the pods i.e. `replicas: 0` - then deploy the pods - increase the chance of success
2. Delete the pods and let the Deployment try to recreate them

Once the cluster is up it seems mostly reliable when you expand the cluster

So the Security Groups for Pods and the VPNs does seem that it could be related to a root cause.

**Comment by jtokuda7** [comment]
@stefan-rink , I changed the wifi network adapter priority and it worked for me! Thank you very much!

**Comment by rehand** [comment]
Same here, changing the metric of the VirtualBox network adapter to a higher value (= lower priority) than VPN works.

Also when H2 is used providing  parameter `--kc-db-url` with a connection string without `AUTO_SERVER=TRUE;` works.

**Comment by vksaamy** [comment]
In my setup, we have many services running in Tanzu platform and all the services connect through mysql router. All services run without any JDBC acquisition connection timeout issue but Keycloak have this issue intermittently. Sometimes the deployment have many restarts due to this issue. When I use JMeter to do a load test with 100 threads, most of the runs have failure rate of 10 to 30% due to this error. Sometimes all request pass and the failure rate is 0%. This error is quite erratic. I already have below config for Keycloak:

    kcDb: mysql
    kcDbPoolInitialSize: "200"
    kcDbPoolMaxSize: "200"
    kcDbPoolMinSize: "200"
    kcDbUrl: jdbc:mysql://mysql-router.abc-ding11.svc.cluster.local:3306/keycloak?enabledTLSProtocols=TLSv1.2 

Tried adding quakus timeout in quarkus.conf file mounted as volume into Keycloak PODs. Seems like don't take effect.

This is a pressing issue for me now as there is an imminent load test to pass for deployment.

Any help on fixing the issue? Thanks in advance.
